### PR TITLE
Add DCGM Monitor

### PR DESCRIPTION
**Pull request by Tabrizian**
The files below are copied from the official DCGM Python 2 bindings. I made some modifications to the documentation, styling, and errors because of using Python 2 and to be consistent with the style that we are using for our code.

* `dcgm_agent.py`
* `dcgm_field_helpers.py`
* `dcgm_fields.py`
* `dcgm_structs.py`
* `dcgm_value.py`

### CODE REVIEW

File: model_analyzer/monitor/dcgm/dcgm_agent.py [hunk 1]
```diff
@@ -0,0 +1,839 @@
...
+# CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+# EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+# PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+# PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+# OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+import model_analyzer.monitor.dcgm.dcgm_structs as dcgm_structs
+from ctypes import *
+# Provides access to functions from dcgm_agent_internal
```
Comment (dzier) [inline model_analyzer/monitor/dcgm/dcgm_agent.py:28]: Empty line before comment

File: model_analyzer/monitor/dcgm/dcgm_agent.py [hunk 2]
```diff
@@ -0,0 +1,839 @@
...
+dcgmFP = dcgm_structs._dcgmGetFunctionPointer
+
+
+# This method is used to initialize DCGM
+def dcgmInit():
```
Comment (dzier) [inline model_analyzer/monitor/dcgm/dcgm_agent.py:33]: Is there a reason these are global methods? Can they be wrapped in a class or namespace? Also, the function name is inconsistent with the rest of the code base. 

1. Method should be `use_underscores_and_lower_case_text`
2. Should have a readable verb, such as `init_dcgm()`, or better yet `dcgm.init()`
3. No method, variable, or class should have a type in their name, i.e. no Hungarian notation.

File: model_analyzer/monitor/dcgm/model.py [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+import model_analyzer.monitor.dcgm.dcgm_agent as dcgm_agent
+import model_analyzer.monitor.dcgm.dcgm_fields as dcgm_fields
+import model_analyzer.monitor.dcgm.dcgm_field_helpers as dcgm_field_helpers
+import model_analyzer.monitor.dcgm.dcgm_structs as structs
+
+from multiprocessing.pool import ThreadPool
+import time
+
+
+class DCGMMonitor(Monitor):
```
Comment (dzier) [inline model_analyzer/monitor/dcgm/model.py:38]: File should be named after class: `dcgm_monitor.py`

**Review by dzier — CHANGES_REQUESTED** [review]

File: model_analyzer/monitor/dcgm/model.py [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+    """
+    Use DCGM to monitor GPU metrics
+    """
+
+    def __init__(self, frequency, dcgmPath=None):
+        """
+        Parameters
+        ----------
+        frequency : int
+            Sampling frequency for the metric
```
Comment (xprotobeast2) [inline model_analyzer/monitor/dcgm/model.py:48]: missing parameter hint for `dcgmPath`

File: model_analyzer/monitor/dcgm/model.py [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+                num_metrics = len(
+                    metrics[dcgm_fields.DCGM_FI_DEV_FB_FREE].values)
+                for i in range(num_metrics):
+                    free_memory = metrics[
+                        dcgm_fields.DCGM_FI_DEV_FB_FREE].values[i]
+                    used_memory = metrics[
+                        dcgm_fields.DCGM_FI_DEV_FB_USED].values[i]
+                    total_memory = metrics[
+                        dcgm_fields.DCGM_FI_DEV_FB_TOTAL].values[i]
+                    memory_record = GPUMemoryRecord(gpu, used_memory.value,
```
Comment (xprotobeast2) [inline model_analyzer/monitor/dcgm/model.py:133]: Merging [this PR](https://github.com/triton-inference-server/model_analyzer/pull/14) will break this. Not sure how cases like this are generally handled. Would you rebase on my branch, or have mine merged first then rebase? In any case this needs to be changed to fit the new format.

File: model_analyzer/monitor/dcgm/model.py [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+                                                    total_memory.value,
+                                                    free_memory.value)
+                    record_collector.insert(memory_record)
+        return record_collector
```
Comment (xprotobeast2) [inline model_analyzer/monitor/dcgm/model.py:137]: New format, doesn't return record collector but just a list of records.

File: model_analyzer/monitor/dcgm/model.py [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+
+    def stop_recording_metrics(self):
+        """Stop recording metrics
+
+        Returns
+        -------
+        RecordCollector
+            A RecordCollector containing all the results
```
Comment (xprotobeast2) [inline model_analyzer/monitor/dcgm/model.py:145]: Here as well

File: model_analyzer/monitor/dcgm/model.py [hunk 8]
```diff
@@ -0,0 +1,158 @@
...
+# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+# PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE COPYRIGHT OWNER OR
+# CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+# EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+# PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+# PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+# OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+from model_analyzer.record.gpu_memory_record import GPUMemoryRecord
```
Comment (xprotobeast2) [inline model_analyzer/monitor/dcgm/model.py:26]: Leave a line here

File: tests/test_dcgm_monitor.py [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+# PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE COPYRIGHT OWNER OR
+# CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+# EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+# PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+# PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+# OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+import unittest
```
Comment (xprotobeast2) [inline tests/test_dcgm_monitor.py:26]: Leave a line

File: tests/test_dcgm_monitor.py [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+        dcgm_monitor = DCGMMonitor(frequency)
+        dcgm_monitor.start_recording_metrics(['memory'])
+        time.sleep(monitoring_time)
+        metrics = dcgm_monitor.stop_recording_metrics()
+        dcgm_monitor.destroy()
+
+        # Assert instance types
+        for i in range(metrics.size()):
+            metric = metrics.get(i)
+            self.assertIsInstance(metric, GPUMemoryRecord)
```
Comment (xprotobeast2) [inline tests/test_dcgm_monitor.py:50]: In [this PR](https://github.com/triton-inference-server/model_analyzer/pull/14) I changed the test to check not only the instances but that there were the same number in the resulting list. Let me know if you think we should do something similar here.

**Review by xprotobeast2 — CHANGES_REQUESTED** [review]

File: model_analyzer/monitor/dcgm/model.py (hunk 5 above)
Comment (Tabrizian) [reply model_analyzer/monitor/dcgm/model.py:133]: I can wait until your PR is merged and then rebase.

**Review by xprotobeast2 — APPROVED** [review]

File: model_analyzer/monitor/dcgm/dcgm_monitor.py [hunk 11]
```diff
@@ -0,0 +1,128 @@
...
+            dcgm_handle, fields, 'triton-monitor')
+
+        self.group_watcher = dcgm_field_helpers.DcgmFieldGroupWatcher(
+            dcgm_handle, self.group_id, self.dcgm_field_group_id.value,
+            structs.DCGM_OPERATION_MODE_MANUAL, frequency, 3600, 0, 0)
+
+    def _monitoring_iteration(self):
+        self.group_watcher.GetMore()
+
+    def _collect_records(self):
```
Comment (xprotobeast2) [inline model_analyzer/monitor/dcgm/dcgm_monitor.py:102]: I think we're including a short docstring for private functions

Comment (Tabrizian) [reply model_analyzer/monitor/dcgm/dcgm_monitor.py:102]: I added a description in the parent class. I will add it here too.

**Review by dzier — APPROVED** [review]

**Pull request merged**